### PR TITLE
[Merged by Bors] - feat: `Fintype.balance` and complex numbers

### PR DESCRIPTION
**Pull request by YaelDillies**
From LeanAPAP

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 1]
```diff
@@ -239,6 +241,13 @@ instance (priority := 100) charZero_rclike : CharZero K :=
 lemma ofReal_expect {α : Type*} (s : Finset α) (f : α → ℝ) : 𝔼 i ∈ s, f i = 𝔼 i ∈ s, (f i : K) :=
   map_expect (algebraMap ..) ..
 
+@[norm_cast]
+lemma ofReal_balance {ι : Type*} [Fintype ι] (f : ι → ℝ) (i : ι) :
+    (↑(balance f i) : K) = balance ((↑) ∘ f) i := map_balance (algebraMap ..) ..
```
Comment (b-mehta) [inline Mathlib/Analysis/RCLike/Basic.lean:246]:
```suggestion
    ((balance f i : ℝ) : K) = balance ((↑) ∘ f) i := map_balance (algebraMap ..) ..
```
this is clearer imo

File: Mathlib/Analysis/RCLike/Basic.lean [hunk 2]
```diff
@@ -239,6 +241,13 @@ instance (priority := 100) charZero_rclike : CharZero K :=
 lemma ofReal_expect {α : Type*} (s : Finset α) (f : α → ℝ) : 𝔼 i ∈ s, f i = 𝔼 i ∈ s, (f i : K) :=
   map_expect (algebraMap ..) ..
 
+@[norm_cast]
+lemma ofReal_balance {ι : Type*} [Fintype ι] (f : ι → ℝ) (i : ι) :
```
Comment (b-mehta) [inline Mathlib/Analysis/RCLike/Basic.lean:245]:
```suggestion
lemma ofReal_balance_apply {ι : Type*} [Fintype ι] (f : ι → ℝ) (i : ι) :
```
optional

**Review by b-mehta — APPROVED** [review]

File: Mathlib/Analysis/RCLike/Basic.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Analysis/RCLike/Basic.lean:245]: Given that you can't take `ofReal` of `balance f` (but you can compose, cf `ofReal_comp_balance`), I'm going to say my name is the better name

File: Mathlib/Data/Complex/BigOperators.lean [hunk 3]
```diff
@@ -28,6 +29,13 @@ theorem ofReal_sum (f : α → ℝ) : ((∑ i ∈ s, f i : ℝ) : ℂ) = ∑ i 
 lemma ofReal_expect (f : α → ℝ) : (𝔼 i ∈ s, f i : ℝ) = 𝔼 i ∈ s, (f i : ℂ) :=
   map_expect ofReal ..
 
+@[simp, norm_cast]
+lemma ofReal_balance [Fintype α] (f : α → ℝ) (a : α) :
+    (↑(balance f a) : ℂ) = balance ((↑) ∘ f) a := by simp [balance]
```
Comment (b-mehta) [inline Mathlib/Data/Complex/BigOperators.lean:34]: as before

Comment (YaelDillies) [reply Mathlib/Data/Complex/BigOperators.lean:34]:
```suggestion
    ((balance f a : ℝ) : ℂ) = balance ((↑) ∘ f) a := by simp [balance]
```

**Review by b-mehta — COMMENTED** [review]
Thanks!

bors merge

**Pull request closed without merging**